### ISSUE TITLE
Third party libraries

### ISSUE THREAD

**Issue by mageshbala90** [asadsahi/AspNetCoreSpa#83]
Hi Asadsahi,

can you please advice me how should i refer third party libraries like primeng in this project?

and also i would like to prevent it from compilation on each build...

I tried to add it in webpack.dev.js file - vendor array, but it does not work properly..  Nodeservices were throwing some exceptions whenever i add it.

**Comment by asadsahi** [comment]
@mageshbala90 try these steps:

1. npm i primeng --save
2. import primeng necessary module in AppModule or SharedModule (you need to export those modules from shared modules to be able to use it in another module)
e.g
import {AccordionModule} from 'primeng/components/accordion/accordion';
![image](https://cloud.githubusercontent.com/assets/3470892/25476575/87377bf2-2b32-11e7-8cb0-1e3a10ebba5a.png)

NB: I haven't installed primeng, hence getting red lines, but you get the idea.

3. Import primeng css files in vendor.css
![image](https://cloud.githubusercontent.com/assets/3470892/25476658/c7cbdd34-2b32-11e7-9872-c8e0df0b7585.png)

You might not need bootstrap, if you have decided to use primeng themes etc. But check on primeng website if they have compatability with bootstrap 4.

**Comment by mageshbala90** [comment]
Hi @asadsahi ,

Thank you for your advice.  I could integrate it by following above mentioned steps and it works.

But after adding Primeng modules, web pack is rebuilding infinitely even though i made no changes in the code.  Please be noted that there is no error in console.  

<img width="960" alt="2017-04-27" src="https://cloud.githubusercontent.com/assets/21171864/25492516/d47f6762-2ba5-11e7-9613-4b73447296a7.png">

**Comment by kiwiingenuity** [comment]
An observation that may be of assistance.
I have not added any third party libraries.
However since the last update(s) my application randomly starts building infinitely.

**Comment by m2017atTR** [comment]
Same problem i had, 
and sawy that happening when some code in ts files has build error. I suggest to look output carefully and clear warning as much as possible.
I tryed with PrimeNg,Material2 and working well !

**Comment by asadsahi** [comment]
@m2017atTR @mageshbala90 I have just trid with two changes in home.component.ts and can't see it re-building more than once.

![image](https://cloud.githubusercontent.com/assets/3470892/25657508/1a66322a-2ff6-11e7-8ceb-63c4206769c7.png)

Can you notice anything specific on your environment?

**Comment by m2017atTR** [comment]
I ment that there can be some changes in ts file changes in local copy 
so when added third party HMR reflects repeatedly build.

**Comment by MaklaCof** [comment]
I am using Kendo UI for Angular, but importing **ButtonsModule** and **LayoutModule** in shared.module.ts or app.module.ts is not working. I had to import to home.module.ts.

shared.module.ts:
```
//.........
// Services
import { ContentService } from './services/content.service';

import { ButtonsModule } from '@progress/kendo-angular-buttons';   //<-------------------
import { LayoutModule } from '@progress/kendo-angular-layout';       //<-------------------


@NgModule({
    imports: [
        CommonModule,
        FormsModule,
        ReactiveFormsModule,
        RouterModule,
        TranslateModule,
        NgbModule.forRoot(),
        // No need to export as these modules don't expose any components/directive etc'
        HttpModule,
        JsonpModule,
        ButtonsModule,                                  //<-------------------
        LayoutModule                                    //<-------------------
    ],
    declarations: [
        DynamicFormComponent,
        DynamicFormControlComponent,
        ErrorMessageComponent,
        ErrorSummaryComponent,
        FooterComponent,
        HeaderComponent,
        PageHeadingComponent,
        UppercasePipe
    ],
    exports: [
        // Modules
        CommonModule,
        FormsModule,
        ReactiveFormsModule,
        RouterModule,
        NgbModule,
        TranslateModule,
        // Providers, Components, directive, pipes
        DynamicFormComponent,
        DynamicFormControlComponent,
        ErrorSummaryComponent,
        ErrorMessageComponent,
        FooterComponent,
        HeaderComponent,
        PageHeadingComponent,
        UppercasePipe,
        ButtonsModule,                                  //<-------------------
        LayoutModule                                    //<-------------------
    ]

})
export class SharedModule
{
    public static forRoot(): ModuleWithProviders
    {
        return {
            ngModule: SharedModule,
            providers: [
                FormControlService,
                ContentService
            ]
        };
    }
}
```

home.component.html:
```
//......
<button kendoButton (click)="onButtonClick()" [primary]="true">My Kendo UI Button</button>
```

Error: 

> Can't bind to 'primary' since it isn't a known property of 'button'.

What am I doing wrong? (_If I import in home.module.ts it works._)

**Comment by asadsahi** [comment]
@mageshbala90 you import/export in sharedmodule looks good. You need to import Shared module in each of consuming module e.g HomeModule in your case.

**Comment by valentinmarinro** [comment]
Hi @asadsahi ,

I have the same issue as @mageshbala90 and @m2017atTR  after installing primeng HMR is rebuilding infinitely, also checked the output and console and do not see any errors / exceptions

![image](https://cloud.githubusercontent.com/assets/1761025/25775755/66878070-32b5-11e7-91a0-0e16c1efe5f1.png)

**Comment by m2017atTR** [comment]
Hello friend,
I could add many other modules and working well...
Faced with same issues then 
- deleted dist folder completely then rebuild system...
- also think about css or scss can couse this problem too... revert or completely remove from project for testing problem.

**Comment by m2017atTR** [comment]
Update  about subject;

today I am working at home, same project at office HMR working perfect but at home
continuously building project as described. so problem result of some global npm or ... packages.

fyi

**Comment by asadsahi** [comment]
@m2017atTR if you have done a production build of this project this creates main.[hash].js file. and then if you run the project in dev mode, it may be reading the hashed version of main.js. Try deleting the wwwroot/dist folder and then run the project. That is the only reason I can think of right now for not running project under HMR.

**Comment by asadsahi** [comment]
I am still trying to find appropriate solutions of this problem, but for now try this workaround.

**Comment by devchild** [comment]
Hi,

I have a related problem.
I am trying to integrate a custom theme, but I am stuck.
I added to theme files in the wwwroot\assets directory.
I imported the css files from the theme using import('path to theme css')

When I run the application, I get an error that the font used in the theme can't be resolved. The font is also in the assets directory, but the theme can't resolve it as the css gets moved into the dist folder but not the font. Can anybody help? My guess is I need to add something in the webpack config but I just can't figure it out.
